### PR TITLE
Round1

### PR DESCRIPTION
**Pull request by yangjunqingqiao**
Initial code refactor, creating template for unit tests, automation, and code style.

### CODE REVIEW

**Review by dnsosa — COMMENTED** [review]
Do we want to delete these Sphinx files? I believe they might be useful later but @cthoyt can perhaps comment.

Regarding the questions about length and number of spaces limitations, these may be arbitrary for quick processing, but this would require checking in with Ben.

File: main_pipeline/tests/util_tests.py
Comment (cthoyt) [inline main_pipeline/tests/util_tests.py]: would be better to reorganize as a python package rather than doing hacky stuff like this. I could make a PR to this PR, or you could hang tight until my Python Packaging talk next week :)

**Comment by cthoyt** [comment]
> Do we want to delete these Sphinx files? I believe they might be useful later but @cthoyt can perhaps comment.
> 
> Regarding the questions about length and number of spaces limitations, these may be arbitrary for quick processing, but this would require checking in with Ben.

There should only be one `docs/` folder per repo, and it should be at the top level. Since the one I PR'd is still there, then any other sphinx docs should be migrated there, and these can be deleted

File: main_pipeline/tests/util_tests.py
Comment (yangjunqingqiao) [reply main_pipeline/tests/util_tests.py]: Yeah, I thought it looked kind of sketchy when I added this. I'll wait on your presentation next week and update accordingly; I'm making this PR a draft in the meantime.

Comment (dnsosa) [reply main_pipeline/tests/util_tests.py]: Actually @cthoyt if this doesn't take too much work would you mind getting us started with reorganizing as a python package? Otherwise we can get started on Wednesday, but I figure might as well get a head start if we can

Comment (cthoyt) [reply main_pipeline/tests/util_tests.py]: What we've got so far is the `src/coronawhy_vt/` directory, where all code must go to be accessible as a package. This is already set up, ready to go, and documented.

I will make a PR for using tox and flake8 for ensuring code quality as we start adding code in here. Because I assume most code will be abandoned after its first use, so we better ensure due diligence is done on inclusion 👅

Comment (cthoyt) [reply main_pipeline/tests/util_tests.py]: @dnsosa ask and you shall receive... but this is going to cause a lot of friction. In my experience, people will do whatever they possibly can to avoid conforming to rules for code quality. Anyway, see: https://github.com/CoronaWhy/task-vt/pull/44

File: .travis.yml
Comment (cthoyt) [inline .travis.yml]: this should be taken care of programmatically with python. Use ` from urllib.request import urlretrieve` for similar functionality inside your code

File: .vscode/settings.json
Comment (cthoyt) [inline .vscode/settings.json]: best not to commit personal project settings

File: src/coronawhy_vt/utils/tf_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/tf_processing.py]: switch to relative imports

Comment (yangjunqingqiao) [reply src/coronawhy_vt/utils/tf_processing.py]: Temporarily commented these out since they aren't used in the current iteration of the functions. Will come back to this when we readd them.

File: .vscode/settings.json
Comment (yangjunqingqiao) [reply .vscode/settings.json]: Done!

**Comment by yangjunqingqiao** [comment]
Not sure where I should put new tickets but TODO is to create a mock of required spaCy models so that tests don't take so long to run.

File: src/relation_extraction_model/model_tests.py
Comment (cthoyt) [inline src/relation_extraction_model/model_tests.py]: use relative imports

File: src/relation_extraction_model/model_tests.py
Comment (cthoyt) [inline src/relation_extraction_model/model_tests.py]: all of the `relation_extraction_model/` files should come inside `src/coronawhy_vt/`

File: src/relation_extraction_model/model_tests.py
Comment (yangjunqingqiao) [reply src/relation_extraction_model/model_tests.py]: Ah, I'll take care of these as the refactoring goes on. Right now the refactored code is only in src/ everything else was a result of merging main into the Round1 branch to grab the stuff from your packaging PR.

Comment (cthoyt) [reply src/relation_extraction_model/model_tests.py]: ahh okay! all the more reason to get the rebase in order :)

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: I guess none of the style checkers caught this, but this R-style syntax isn't appropriate in python. It would look best with a space after `return` and no unnecessary parentheses like below:

```python
return [{'text': paper['metadata']['title']}]
```

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py:87]: Is this function really necessary? It adds 9 lines where it might just be more appropriate to write `paper['body_text']` anywhere you would call it. This philosophy comes from Jack Diedrich (see his excellent talk at https://www.youtube.com/watch?v=o9pEzgHorH0)

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: c-style iteration could be replaced with a beautiful list comprehension like this:

```python
words = [
    word.lower()
    for word in words
]
```

More on list comprehensions from David Beazley in https://www.youtube.com/watch?v=j6VSAsKAj98

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py:194]: big problem if you're using local paths - this means not a single other person will ever be able to use this code

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: this is c-style syntax as well, would look better as 

```python
if lemmatized:
```

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: more readable if you do tuple unpacking here and assign meaningful names to `i[0]` and `i[1]` like in :

```python
for variable_1_name, variable_2_name in index:
    index_1.append(variable_1_name)
    index_2.append(variable_2_name)
```

File: src/coronawhy_vt/utils/local_processing.py
Comment (yangjunqingqiao) [reply src/coronawhy_vt/utils/local_processing.py:194]: This is a placeholder path variable until we get dataverse setup properly. But I agree we have a lot of csvs in the repo itself and I'm going to weed them all out by the time we finish the pipeline.

File: src/coronawhy_vt/utils/local_processing.py
Comment (yangjunqingqiao) [reply src/coronawhy_vt/utils/local_processing.py:87]: Yeah, my thought too although I wanted this PR to just be a move of the original functions into proper package and reproducible test form. I'll open a new PR for streamlining the methods.

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py:299]: this is pretty confusing between the cryptic variable names and the two levels of indexing. can this be improved?

File: src/relation_extraction_model/model_tests.py
Comment (yangjunqingqiao) [reply src/relation_extraction_model/model_tests.py]: resolved due to rebase.

File: src/coronawhy_vt/utils/tf_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/tf_processing.py]: `os.makedirs(dirname, exist_ok=True)` does this functionality exactly

File: tests/__init__.py
Comment (cthoyt) [inline tests/__init__.py]: should be a trailing newline here...

File: tests/constants.py
Comment (cthoyt) [inline tests/constants.py:7]: should check to make sure that the tests are being checked by the flake8 script - there are a lot of parts here that should be caught

File: tox.ini
Comment (cthoyt) [inline tox.ini]: it's cheating if you ignore the problems instead of solving them!

Comment (cthoyt) [reply tox.ini]: or at least thinking about why the problems are happening and applying appropriate `noqa` tags as in the case of S403, which is about unsafe usage of pickle iirc

Comment (yangjunqingqiao) [reply tox.ini]: Those were warnings for pickle being a security vulnerability, I'm going to do some research and implement an alternative in a later PR.

Comment (cthoyt) [reply tox.ini]: there might not be an alternative to using pickle for storage of models, so this is a valid thing to ignore. However, there's a `.flake8` config file where this would be more appropriate instead of `tox.ini`

**Comment by cthoyt** [comment]
@dnsosa would it be possible to make me an admin for this repo so I can make pushes to @yangjunqingqiao's fork and solve the rebase vs master problem? Otherwise I'd have to fork his fork and PR into his fork, wait for him to rebase against it, and possibly end up in a catch-22

As it is, there are 50 files in the diff and it's not possible to see what's actually changed

**Comment by yangjunqingqiao** [comment]
> 
> 
> @dnsosa would it be possible to make me an admin for this repo so I can make pushes to @yangjunqingqiao's fork and solve the rebase vs master problem? Otherwise I'd have to fork his fork and PR into his fork, wait for him to rebase against it, and possibly end up in a catch-22
> 
> As it is, there are 50 files in the diff and it's not possible to see what's actually changed

@cthoyt The most recent [commit](https://github.com/CoronaWhy/task-vt/pull/41/commits/fcb4cf0f55ff418a6ef21623bbaf2f20bd8d11d9) should show just the files that I changed with regard to your previous code review comments (10 files changed). I'm not too familiar with code reviews in github so let me know if that's not the view for you and I can try cleaning up the branch again!

**Comment by cthoyt** [comment]
When we're judging a PR, we shouldn't look at individual commits, but rather the whole PR. The best practice is: before merge to squash all of the commits together into one. This allows people to make whatever changes they'd like during the development process whether they're small or large.

The reason I keep bringing up a proper rebase vs master is this number here:

Screen Shot 2020-06-13 at 20 42 41

When you [click this tab](https://github.com/CoronaWhy/task-vt/pull/41/files) to see what changes are being made by this PR, you see nonsense (copying the changes made in other commits). We can't do a good job reviewing with this because it's not only distracting, but makes it impossible to tell what this PR would actually do if changed. Please let me know if this still isn't clear (there are tons of articles about rebasing vs. master. if you're having trouble finding one with google let me know and I can pick one for you)

**Comment by dnsosa** [comment]
@cthoyt I've added you to the repo with Maintainer privileges, which is what I've given to all the project leads. Will pushing to @yangjunqingqiao's fork require full Admin privileges or should Maintainer suffice?

**Comment by yangjunqingqiao** [comment]
@cthoyt I finally understood what you meant by your first comment. Instead of rebase I created a new branch and squash merged with master back onto the original Round1 branch. This commit also contains the changes I commented on your new PR. 

As a side note, is it better practice after doing a squash to open a new PR or have it appended to the messy old one?

**Comment by cthoyt** [comment]
@yangjunqingqiao well if there's a history of comments and discussion then it's generally bad to start a new PR. for the purposes of reviews, it's also nice to have the commit history. However, github has the "squash and merge" option which is best for project management (so squashing gets done after everyone is happy with the code)

So what will you do next?

**Comment by yangjunqingqiao** [comment]
@cthoyt In that case then I think it would be best to have a final review on the last squash commit I made and push this PR after everyone is happy with it. I also added the changes that I commented on your round1-reset branch.

File: .readthedocs.yml
Comment (cthoyt) [inline .readthedocs.yml]: @yangjunqingqiao if the rebase has been done, why is this still in the files changed list?

Comment (yangjunqingqiao) [reply .readthedocs.yml]: Ah, good catch I must’ve rebased wrt to a version of main before your packaging PR was merged. I’ll update it real fast.

**Comment by yangjunqingqiao** [comment]
I'm definitely not doing something right, I've already rebased wrt to master and squashed the commits from the branch but it's still showing the old files changed. I need a bit of time to figure out what's going on

**Comment by yangjunqingqiao** [comment]
Figured it out, according to [this](https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch), the target branch determines the diffs seen in the PR screen. Since Round1 wasn't up to date with main before the PR no amount of rebasing would fix the issue as the upstream Round1 branch does not have those changes. 

The solution is to change the target branch to main to filter out files that were not modified for this specific PR (I tested this with @cthoyt 's last rebasing PR by swapping his target to Round1 and it also still tracks the commits already made to main).

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: use a context manager around the file to automatically close it

```python
with open(...) as file:
   ...
```

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: iterating over lines in a file is good enough - you don't have to do file.readlines()

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: make a new line after `replace_dict: dict = None` but before the closing parentheses `)` to make this a bit more readable

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: use a context manager to automatically close the file when you're done
```python
with (path / p).open() as file:
    paper = json.load(readfile)
```

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: remove excess parentheses

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: if you use `enumerate()` you get the index of the block and the block itself at the same time, making it a lot easier to read the code later
```python
for b, block in enumerate(blocks):
    ....
                            block_text.append(block['text'])
                            block_id.append(b)
```

`block` is the same as `blocks[b]`

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: remove redundant parentheses

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: since you're going two levels deep in independent iterators, it's better to use `itertools`

```python
import itertools as itt
for t, w in itt.product(text, tokenized_words):
    ...
```

don't worry - the iteration will go in exactly the same order

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: I like dicts better! Much more self-documentation.

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py:127]: there's lots of overlap in this function versus the last one. Maybe there's the possibility to abstract?

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: what happens if `data_dicts` has a length of greater than 1? Everything after the first element gets thrown out?

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: It's not so obvious what's going on here, could you add some comments?

Maybe this functionality could be extracted and put in its own function, where it would be easier to document what it's doing and why

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: Could be written more idiomatically with usage of the `any()`
```python
return not any(
    s in interstring
    for s in sentence enders
)
```

or apply [De Morgan's law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) to use `all()`

```python
return all(
    s not in interstring
    for s in sentence enders
)
```

Comment (cthoyt) [reply src/coronawhy_vt/utils/local_processing.py]: Not that it's a big deal for efficiency because it's a list of length 4, but the second one allows for better short-circuit evaluation

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: there's a totally mindblowing pattern with `zip()` for this that lets you do it in one line:

```python
index_1, index_2 = zip(*index)
```

File: tests/test_utils.py
Comment (cthoyt) [inline tests/test_utils.py]: better style to have a newline after the open bracket `[` and at the end before the close bracket `]`

File: tests/test_utils.py
Comment (cthoyt) [inline tests/test_utils.py]: pro mode - you don't need a pass if there's a docstring :)

File: tests/test_utils.py
Comment (cthoyt) [inline tests/test_utils.py]: Don't add this `if __name__ == '__main__'` section at all - running the tests should be the job of `tox`

File: tests/test_utils_log.txt
Comment (cthoyt) [inline tests/test_utils_log.txt]: Don't commit logs

File: utils/data_utils_test.py
Comment (cthoyt) [inline utils/data_utils_test.py]: best not to use relative directory paths at all in the code. you can use the constants module to build the file path in a more reusable way

File: utils/data_utils_test.py
Comment (cthoyt) [inline utils/data_utils_test.py]: remove this - `tox` decides how the tests are run

File: tests/test_utils.py
Comment (cthoyt) [inline tests/test_utils.py]: if you put parentheses around the imports, you don't need ugly line breaks. A modern IDE like PyCharm should automatically reformat this stuff for you

File: setup.py
Comment (cthoyt) [inline setup.py]: let's remove this change and minimize our diff

File: src/coronawhy_vt/utils/local_processing.py
Comment (yangjunqingqiao) [reply src/coronawhy_vt/utils/local_processing.py]: I agree, however in this case the dict values represent individual lists which then have to be aggregated to get entries for similarities. i.e. 

`mentions = pull_mentions_lemmatized()`
`first_entry = (mentions['sha'][0], mentions['blockid'][0], mentions['word'][0], mentions['sec'][0], mentions['pos'][0], mentions['block'][0])`

whereas if we used a list of tuples (or even a list of dicts)

`mentions = pull_mentions_lemmatized`
`first_entry = mentions[0]`

Seems like it would be cleaner to use for the end user.

File: src/coronawhy_vt/utils/local_processing.py
Comment (yangjunqingqiao) [reply src/coronawhy_vt/utils/local_processing.py:127]: Considering moving some of these preprocessing functions into classes (sort of like how streamed tokenizers are done for spaCy and TensorFlow).

Probably an issue for a later PR.

Comment (cthoyt) [reply src/coronawhy_vt/utils/local_processing.py:127]: In my experience if it doesn't get done when it's point out the first time, it will never get done. I'd appreciate if you take care of it now - it will make it easier for anyone else reviewing the code in the future too

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [reply src/coronawhy_vt/utils/local_processing.py]: Maybe a namedtuple would be best then if the list is always the same size

File: src/coronawhy_vt/utils/local_processing.py
Comment (cthoyt) [inline src/coronawhy_vt/utils/local_processing.py]: if you make the `Feature` class module-level instead of being inside the function, it might make it easier to do type annotations and for the IDE to make type inference

Comment (yangjunqingqiao) [reply src/coronawhy_vt/utils/local_processing.py]: Would you recommend creating a new file for it or just putting it at the top of local_processing?

Comment (cthoyt) [reply src/coronawhy_vt/utils/local_processing.py]: just at the top of local_processing is fine

File: setup.py
Comment (cthoyt) [reply setup.py]: @yangjunqingqiao please take care of this

**Review by cthoyt — APPROVED** [review]
Stuff's looking a lot better now! I hope everything's still working after the refactor (and that the unittests are enough to check that)

There are 2 tiny comments but after that go ahead and merge (after Travis-CI is passing, of course)

File: tox.ini
Comment (cthoyt) [inline tox.ini]: put `tests/` on the first line so there's only one set of feedback in this run like:

```
commands =
    flake8 src/coronawhy_vt/ tests/ setup.py
```

File: setup.py
Comment (yangjunqingqiao) [reply setup.py]: I'm confused what should happen for this, Flake8 will complain if I revert this change (no ending newline). Should I ignore the specific error or is there a way to untrack the change in the PR?

Comment (cthoyt) [reply setup.py]: It's just that there should not be a space between the line `if __name__ == '__main__':` and `setuptools.setup`. There should be an ending newline, which is unrelated to this line of diff

**Comment by cthoyt** [comment]
@yangjunqingqiao  looks like there's a problem with the automated checks - it's looking for there to be a "branch". Maybe you have to activate Travis-CI on your personal GitHub account? Try going to travis-ci.com and signing up, and see what it looks like at https://travis-ci.com/github/yangjunqingqiao/task-vt. Otherwise, we can ask @dnsosa to merge with admin powers

**Comment by yangjunqingqiao** [comment]
Currently trying to debug, apparently it's a known [issue](https://travis-ci.community/t/travis-ci-branch-expected-waiting-for-status-to-be-reported/829).

I have Travis CI enabled on my fork already I think it's an issue with the branch protection settings on main. Going to try switching the target to Round1 and see if that passes.

**Comment by yangjunqingqiao** [comment]
Weird, Travis CI is setup for the fork but my fork doesn't seem to be sending the success message to the upstream.

**Comment by yangjunqingqiao** [comment]
@dnsosa I think we might have to just override the branch check. It seems that it's currently an unresolved recurring [issue](https://travis-ci.com/yangjunqingqiao/task-vt.svg?branch=Round1) with travis-ci.

**Comment by cthoyt** [comment]
@yangjunqingqiao 🎊🎊🎊🎊 it's done :)

**Comment by yangjunqingqiao** [comment]
@cthoyt Thanks for all the feedback and help, I learned a lot!

**Pull request merged**
